grails-app/services/org/pih/warehouse/data/PersonService.groovy

```diff
@@ -11,33 +11,136 @@ package org.pih.warehouse.data
+
+    /**
+     * @param recipient Either a RFC822O internet/email address (ex: "Justin Miranda <justin@openboxes.com>")
+     *        or simply the recipient's name (ex: "Justin Miranda"). We accept both formats since either
+     *        format can be provided during data imports.
+     * @return A Person who has the provided name and email. Will be a new Person if they didn't previously exist.
+     */
+    Person getOrCreatePersonByRecipient(String recipient) {
+        if (StringUtils.isBlank(recipient)) {
+            return null
+        }
+
+        InternetAddress internetAddress
+        try
+        {
+            internetAddress = new InternetAddress(recipient, false)
+        }
+        catch (AddressException ignored) {
+            // If recipient isn't a valid internet address, it must be a regular name.
+            return getOrCreatePersonFromNames(recipient)
+        }
+
+        // If a person exists with the given email, return them. Note that this can return inactive people so
+        // the caller must properly handle that case. We do this to avoid creating duplicate users.
+        Person person = getPersonByEmail(internetAddress.address)
+        if (person) {
+            return person
+        }
+
+        // Otherwise create a new person.
+        if (!internetAddress.personal) {
+            throw new RuntimeException("Cannot save new recipient without a name: ${recipient}")
+        }
+        String[] names = extractNames(internetAddress.personal)
+        person = new Person(firstName: names[0], lastName: names[1], email: internetAddress.address)
+        if (!person.save(flush: true)) {
+            throw new ValidationException("Cannot save recipient ${recipient} due to errors", person.errors)
+        }
+        return person
     }
 
-    Person getOrCreatePersonFromNames(String combinedNames) {
+    /**
+     * @param combinedNames The person's full name (first name + last name) combined into a single String.
+     * @return A Person who has the provided combinedName. Will be a new Person if they didn't previously exist.
+     */
+    private Person getOrCreatePersonFromNames(String combinedNames) {
+        // If a person exists with the given name, return them. Note that this can return inactive people so
+        // the caller must properly handle that case. We do this to avoid creating duplicate users.
+        Person person = getPersonByName(combinedNames)
+        if (person) {
+            return person
+        }
+
         String[] names = extractNames(combinedNames)
-        Person person = getPersonByNames(names)
-        if (!person) {
-            person = new Person(firstName: names[0], lastName: names[1])
-            if (!person.save(flush: true)) {
-                throw new ValidationException("Cannot save recipient ${combinedNames} due to errors", person.errors)
-            }
+        person = new Person(firstName: names[0], lastName: names[1])
+        if (!person.save(flush: true)) {
```

Comment:
and this one flush?

Reply:
Same answer. It was in the original code so I didn't want to change it without knowing why it was added